sklearn/neighbors/_nca.py

```diff
@@ -216,6 +218,7 @@ def fit(self, X, y):
         # Verify inputs X and y and NCA parameters, and transform a copy if
         # needed
         X, y, init = self._validate_params(X, y)
+        self._n_features_out = self.n_components or X.shape[1]
```

Comment:
I think it's more maintainable to get the feature names out directly from `components_`:

https://github.com/scikit-learn/scikit-learn/blob/1d697842a36fbd93e89d7756bf899f527baf6323/sklearn/neighbors/_nca.py#L251

where `_n_features_out = components_.shape[0]`, because `components_` is used directly in `transform` to compute the output. This way there is less mental overhead to go from `components_` -> `_n_features_out`.